File: frontend/src/features/digital-twin/api/digital_twin.ts
import { api } from "@/lib/api/client";

export interface FacilitySummary {
  id: string;
  name: string;
  location: string | null;
  asset_count: number;
  critical_count: number;
  health_score: number;
  risk_score: number;
  active_incident_count: number;
}

export interface DigitalTwinNode {
  id: string;
  label: string;
  type: string;
  status: string;
  health_score: number;
  risk_score: number;
  risk_level: string;
  location: string | null;
  facility_id: string | null;
  parent_id: string | null;
  active_incidents: number;
  incident_severities: string[];
  maintenance_due: boolean;
  children_count: number;
}

export interface DigitalTwinEdge {
  source: string;
  target: string;
  relationship: string;
  label: string;
  weight: number;
}

export interface DigitalTwinTopology {
  nodes: DigitalTwinNode[];
  edges: DigitalTwinEdge[];
}

export interface DigitalTwinOverview {
  total_assets: number;
  total_facilities: number;
  total_nodes: number;
  total_edges: number;
  critical_count: number;
  at_risk_count: number;
  active_incidents: number;
  average_health: number;
  facilities: FacilitySummary[];
}

export const digitalTwinApi = {
  facilities: () => api.get<FacilitySummary[]>("/api/v1/digital-twin/facilities"),

  facility: (id: string) =>
    api.get<DigitalTwinTopology>(`/api/v1/digital-twin/facilities/${encodeURIComponent(id)}`),

  assets: () => api.get<DigitalTwinTopology>("/api/v1/digital-twin/assets"),

  overview: () => api.get<DigitalTwinOverview>("/api/v1/digital-twin/overview"),
};

File: backend/app/schemas/digital_twin.py
"""Pydantic schemas for the Enterprise Digital Twin."""

from __future__ import annotations

from pydantic import BaseModel, Field


class FacilitySummary(BaseModel):
    id: str
    name: str
    location: str | None = None
    asset_count: int
    critical_count: int
    health_score: float = Field(..., ge=0, le=100)
    risk_score: float = Field(..., ge=0, le=100)
    active_incident_count: int


class DigitalTwinNode(BaseModel):
    id: str
    label: str
    type: str
    status: str
    health_score: float = Field(..., ge=0, le=100)
    risk_score: float = Field(..., ge=0, le=100)
    risk_level: str
    location: str | None = None
    facility_id: str | None = None
    parent_id: str | None = None
    active_incidents: int = 0
    incident_severities: list[str] = []
    maintenance_due: bool = False
    children_count: int = 0


class DigitalTwinEdge(BaseModel):
    source: str
    target: str
    relationship: str
    label: str = ""
    weight: float = 1.0


class DigitalTwinTopology(BaseModel):
    nodes: list[DigitalTwinNode]
    edges: list[DigitalTwinEdge]


class DigitalTwinOverview(BaseModel):
    total_assets: int
    total_facilities: int
    total_nodes: int
    total_edges: int
    critical_count: int
    at_risk_count: int
    active_incidents: int
    average_health: float
    facilities: list[FacilitySummary]
